zeroex/orderwatch/order_watcher_test.go

```diff
@@ -1622,6 +1623,226 @@ func TestDrainAllBlockEventsChan(t *testing.T) {
+	database, err := db.New(ctx, db.TestOptions())
+	require.NoError(t, err)
+
+	validator, err := ordervalidator.New(
+		&SlowContractCaller{
+			caller:            ethRPCClient,
+			contractCallDelay: time.Second,
+		},
+		constants.TestChainID,
+		ethereumRPCMaxContentLength,
+		ganacheAddresses,
+	)
+	require.NoError(t, err)
+
+	blockWatcher, orderWatcher := setupOrderWatcherWithValidator(ctx, t, ethRPCClient, database, validator)
+	orderEventsChan := make(chan []*zeroex.OrderEvent, 2*orderWatcher.maxOrders)
+	orderWatcher.Subscribe(orderEventsChan)
+
+	// Create a new order
+	signedOrder := scenario.NewSignedTestOrder(t, orderopts.SetupMakerState(true))
+	err = blockWatcher.SyncToLatestBlock()
+	require.NoError(t, err)
+	orderHash, err := signedOrder.ComputeOrderHash()
+	require.NoError(t, err)
+
+	// Cancel the order
+	opts := &bind.TransactOpts{
+		From:   signedOrder.MakerAddress,
+		Signer: scenario.GetTestSignerFn(signedOrder.MakerAddress),
+	}
+	trimmedOrder := signedOrder.Trim()
+	txn, err := exchange.CancelOrder(opts, trimmedOrder)
+	require.NoError(t, err)
+	waitTxnSuccessfullyMined(t, ethClient, txn)
+
+	syncErrChan := make(chan error)
+	validationErrChan := make(chan error)
+	validationResultsChan := make(chan *ordervalidator.ValidationResults)
+
+	go func() {
+		err = blockWatcher.SyncToLatestBlock()
+		syncErrChan <- err
+	}()
+
+	go func() {
+		validationResults, err := orderWatcher.ValidateAndStoreValidOrders(ctx, []*zeroex.SignedOrder{signedOrder}, false, constants.TestChainID)
+		if err != nil {
+			validationErrChan <- err
+			return
+		}
+		validationResultsChan <- validationResults
+	}()
+
+	err = <-syncErrChan
+	require.NoError(t, err)
+
+	select {
+	case err := <-validationErrChan:
+		t.Error(err)
+	case validationResults := <-validationResultsChan:
+		require.Equal(t, len(validationResults.Accepted), 1)
+		assert.Equal(t, len(validationResults.Rejected), 0)
+		orderEvents := waitForOrderEvents(t, orderEventsChan, 1, 4*time.Second)
+		assert.Equal(t, zeroex.ESOrderAdded, orderEvents[0].EndState)
+		assert.Equal(t, orderHash, orderEvents[0].OrderHash)
+	}
+
+	// Add new block events and then check to see if the order has been removed from the blockwatcher
+	latestBlock, err := database.GetLatestMiniHeader()
+	require.NoError(t, err)
+	nextBlock := &types.MiniHeader{
+		Parent:    latestBlock.Hash,
+		Hash:      common.HexToHash("0x1"),
+		Number:    big.NewInt(0).Add(latestBlock.Number, big.NewInt(1)),
+		Timestamp: latestBlock.Timestamp.Add(15 * time.Second),
+	}
+	newBlockEvents := []*blockwatch.Event{
+		{
+			Type:        blockwatch.Added,
+			BlockHeader: nextBlock,
+		},
+	}
+	orderWatcher.blockEventsChan <- newBlockEvents
+
+	// Await canceled event
+	orderEvents := waitForOrderEvents(t, orderEventsChan, 2, 10*time.Second)
+	assert.Equal(t, zeroex.ESOrderCancelled, orderEvents[0].EndState)
+	assert.Equal(t, orderHash, orderEvents[0].OrderHash)
+	// TODO(jalextowle): This event probably shouldn't be fired
```

Comment:
The issue here was that our query in `findOrdersToUnexpire` contained the filter:

```
db.OrderFilter{
    Field: db.OFFillableTakerAssetAmount,
    Kind: db.NotEqual,
    Value: 0,
}
```

Since `types.OrderWithMetadata.FillableTakerAssetAmount` is a `*big.Int`, this would return true even if the fillable taker asset amount was `big.NewInt(0)`. Changing `query.Value` to `big.NewInt(0)` fixed this issue.